nntrainer/dataset/random_data_producers.cpp

```diff
@@ -0,0 +1,167 @@
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <random_data_producers.h>
+
+#include <base_properties.h>
+#include <node_exporter.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+/**
+ * @brief Props containing min value
+ *
+ */
+class PropsMin : public Property<float> {
+public:
+  /**
+   * @brief Construct a new props min object with a default value
+   *
+   * @param value default value
+   */
+  PropsMin(float value = 0.0f) : nntrainer::Property<float>(value) {}
+  static constexpr const char *key = "min"; /**< unique key to access */
+  using prop_tag = float_prop_tag;          /**< property type */
+};
+
+/**
+ * @brief Props containing max value
+ *
+ */
+class PropsMax : public Property<float> {
+public:
+  /**
+   * @brief Construct a new props max object with a default value
+   *
+   * @param value default value
+   */
+  PropsMax(float value = 1.0f) : nntrainer::Property<float>(value) {}
+  static constexpr const char *key = "max"; /**< unique key to access */
+  using prop_tag = float_prop_tag;          /**< property type */
+};
+
+/**
+ * @brief Props containing max value
+ *
+ */
+class PropsDataSize : public Property<unsigned int> {
+public:
+  /**
+   * @brief Construct a new props data size object with a default value
+   *
+   * @param value default value
+   */
+  PropsDataSize(unsigned int value = 512) :
+    nntrainer::Property<unsigned int>(value) {}
+  static constexpr const char *key = "size"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;            /**< property type */
+};
+
+RandomDataOneHotProducer::RandomDataOneHotProducer() :
+  rd_one_hot_props(new Props()) {}
+
+RandomDataOneHotProducer::~RandomDataOneHotProducer() {}
+
+const std::string RandomDataOneHotProducer::getType() const {
+  return RandomDataOneHotProducer::type;
+}
+
+unsigned long long
+RandomDataOneHotProducer::size(const std::vector<TensorDim> &input,
+                               const std::vector<TensorDim> &output) const {
+  return std::get<PropsDataSize>(*rd_one_hot_props).get();
+}
+
+void RandomDataOneHotProducer::setProperty(
+  const std::vector<std::string> &properties) {
+  auto left = loadProperties(properties, *rd_one_hot_props);
+  NNTR_THROW_IF(!left.empty(), std::invalid_argument)
+    << "There is unparsed properties, size: " << left.size();
+}
+
+DataProducer::Gernerator
+RandomDataOneHotProducer::finalize(const std::vector<TensorDim> &input_dims,
+                                   const std::vector<TensorDim> &label_dims) {
+  /** check if the given producer is ready to finalize */
+  auto &[min_, max_, sz] = *rd_one_hot_props;
+
+  NNTR_THROW_IF(std::any_of(label_dims.begin(), label_dims.end(),
+                            [](const TensorDim &dim) {
+                              return dim.channel() != 1 || dim.height() != 1;
```

Comment:
any reason to have this restriction? IMO consider this class to be a generic data producer which can work for super resolution or object detection samples as well than just classifier. Removing this check will do so.

Reply:
My thinking is that there are number of different random data producer (categorical, sparse)... This is like instead 'loss' has one type + mode, each mode is one type. As the classname is `RandomDataOneHotProducer` this restriction is added.

Added in the class brief that
`* @brief RandomDataProducer which generates a onehot vector as a label`

Reply:
Then it will be RandomDataProducer -> RandomSingleLabelDataProducer (as this cannot be used for multi-class classification problems). If you keep the current name, please add in description that only single label is supported.

Reply:
I recommend to keep RandomDataProducer with removing restriction.  If it is SingleLabel, then it is expected to have MultiLabel and also we need to define what is MultiLabel. ( Channel, Height?? )

Reply:
Could you explain why multilabel is not supported?

if you have more then one label, it's fine.
It's returning one-hot vector and it is the name of the class. (`RandomDataOneHotProducer`)

Reply:
> I recommend to keep RandomDataProducer with removing restriction. If it is SingleLabel, then it is expected to have MultiLabel and also we need to define what is MultiLabel. ( Channel, Height?? )

It's already supporting multi label and not restricted to give single label

eg) giving label dims of {1:1:1:5, 1:1:1:7} will return two labels.

I think we can pass a vectors instead of vaguely using a single label.

The class we are dicussing is `RandomDataOneHotProducer` and it's serving it's purpose. If we need another data producer, I think we can add another class (the reason why `random_data_producers.h` not `random_data_producer.h`)

Reply:
also s/`* @brief RandomDataProducer which generates a onehot vector as a label`/`* @brief A family of random data producer which generates a onehot vector as a label`

Reply:
> Could you explain why multilabel is not supported?
> 

What I mean is that we don't need to just adding another method if we support same feature with small change of others.

Reply:
In that case, we will have a mode='something' rather than a type='something', this might be appropriate if we are to have another type of data producer. I think if we have a need for other data producer, we can decide at that time :)

Reply:
> In that case, we will have a mode='something' rather than a type='something', this might be appropriate if we are to have another type of data producer. I think if we have a need for other data producer, we can decide at that time :)

What I mean is, just removing the error checking of channel and height will work like @kparichay said.

Reply:
Even if we delete this, the code is not supporting multi channel, multi height but only onehot for now.
I'll add a todo instead to support this.

Reply:
> Even if we delete this, the code is not supporting multi channel, multi height but only onehot for now.
> I'll add a todo instead to support this.

Even if it is not able to support the multi channel or multi height, we can give more flexibility. Anyway, Let's leave it as in todo then.

Reply:
> Could you explain why multilabel is not supported?

I guess in that case, there are multiple labels for a given input. So, the one hot output would be of dimension like {b,1,multi_labels, on_hot_vector_size} (where multi_labels value is not fixed and is different for each input), but leaving it as a TODO is fine for now :).